urllib3/util/rfc6555.py

```diff
@@ -0,0 +1,223 @@
+
+            # Successfully found a proper connected socket.
+            if self._socket_check(sock):
+                return sock
+
+        return None
+
+    def _create_socket(self, family, socktype, proto, sa):
+        """ Tries it's best to create a socket and register
+        the new socket with the selector. If it fails, no
+        problem, the algorithm continues as usual. """
+        sock = None
+        try:
+            sock = socket.socket(family, socktype, proto)
+            if self._timeout is socket._GLOBAL_DEFAULT_TIMEOUT:
+                self._timeout = sock.gettimeout()
+
+            if self._socket_options:
+                for opt in self._socket_options:
+                    sock.setsockopt(*opt)
+
+            if self._source_address:
+                sock.bind(self._source_address)
+
+            # Set the socket to non-blocking so we can select it.
+            sock.settimeout(0.0)
+
+            error_code = sock.connect_ex(sa)
+            if error_code and error_code not in _ASYNC_ERROR_NUMBERS:
+                self._error = socket.error(error_code)
+            else:
+                self._sockets.append(sock)
+                self._selector.register(sock, EVENT_WRITE)
+        except (socket.error, OSError) as e:
+            self._error = e
+            if sock is not None:
+                try:
+                    sock.close()
+                except (OSError, socket.error):
+                    pass
+
+    def _socket_check(self, sock):
+        """ Checks a socket to see if it is in an
+        error state, if so deal with cleaning it up
+        Returns True if the socket is still good to use. """
+        error_code = sock.getsockopt(socket.SOL_SOCKET,
+                                     socket.SO_ERROR)
+
+        # If the socket is errored, then we need to clean it up.
+        if error_code and error_code not in _ASYNC_ERROR_NUMBERS:
+            self._error = socket.error(error_code)
+            self._selector.unregister(sock)
+            self._sockets.remove(sock)
+            try:
+                sock.close()
+            except (OSError, socket.error):
+                pass
+            return False
+
+        return True
+
+    def _timeout_check(self):
+        """ Checks to see if the algorithm has timed out
+        and if it has, raise a timeout error. """
+        if self._has_timed_out():
+            raise socket.timeout()
+
+    def _has_timed_out(self):
+        """ Checks to see if the algorithm has timed out. """
+        if self._timeout is socket._GLOBAL_DEFAULT_TIMEOUT:
+            return False
+        elif self._timeout is None:
+            return False
+        else:
+            return current_time() > self._start_time + self._timeout
+
+    def _cleanup(self):
+        for sock in self._sockets:
+            try:
+                sock.close()
+            except (OSError, socket.error):
+                pass
+        self._selector.close()
```

Comment:
We should consider turning this object into a context manager and using it that way, to ensure that we always clean up all our sockets and selectors regardless of how we escape the lookup.